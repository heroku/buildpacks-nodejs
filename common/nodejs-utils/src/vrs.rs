use node_semver::{Range, Version as NSVersion};
use serde::{Deserialize, Serialize};
use std::{error::Error, fmt};

#[derive(Debug)]
pub struct VersionError(String);
impl Error for VersionError {}
impl fmt::Display for VersionError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
#[serde(try_from = "String")]
pub struct Version(NSVersion);

impl Version {
    /// Parses a Node.js semver string as a `Version`.
    ///
    /// # Errors
    ///
    /// Invalid Node.js semver strings will return a `VersionError`
    pub fn parse(version: &str) -> Result<Self, VersionError> {
        let trimmed = version.trim();
        match NSVersion::parse(trimmed) {
            Ok(v) => Ok(Version(v)),
            Err(e) => Err(VersionError(format!("{}", e))),
        }
    }

    /// Returns the major version identifier.
    #[must_use]
    pub fn major(&self) -> u64 {
        self.0.major
    }

    /// Returns the minor version identifier.
    #[must_use]
    pub fn minor(&self) -> u64 {
        self.0.minor
    }

    /// Returns the patch version identifier.
    #[must_use]
    pub fn patch(&self) -> u64 {
        self.0.patch
    }
}

impl TryFrom<String> for Version {
    type Error = VersionError;
    fn try_from(val: String) -> Result<Self, Self::Error> {
        Version::parse(&val)
    }
}
impl fmt::Display for Version {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Deserialize, Debug, Clone)]
#[serde(try_from = "String")]
pub struct Requirement(Range);

impl Requirement {
    /// Parses `package.json` version string into a Requirement. Handles
    /// these cases:
    ///
    /// * Any node-semver compatible string
    /// * "latest" as "*"
    /// * "~=" as "="
    ///
    /// # Errors
    ///
    /// Invalid version strings wil return a `VersionErr`
    pub fn parse(requirement: &str) -> Result<Self, VersionError> {
        let trimmed = requirement.trim();
        if requirement == "latest" {
            return Ok(Requirement(Range::any()));
        }
        if trimmed.starts_with("~=") {
            let version = trimmed.replacen('=', "", 1);
            if let Ok(range) = Range::parse(version) {
                return Ok(Requirement(range));
            }
        }
        match Range::parse(trimmed) {
            Ok(range) => Ok(Requirement(range)),
            Err(error) => Err(VersionError(format!("{}", error))),
        }
    }

    #[must_use]
    pub fn any() -> Self {
        Requirement(Range::any())
    }

    #[must_use]
    pub fn satisfies(&self, ver: &Version) -> bool {
        self.0.satisfies(&ver.0)
    }
}

impl TryFrom<String> for Requirement {
    type Error = VersionError;
    fn try_from(val: String) -> Result<Self, Self::Error> {
        Requirement::parse(&val)
    }
}

impl fmt::Display for Requirement {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_handles_latest() {
        let result = Requirement::parse("latest");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!("*", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_exact_versions() {
        let result = Requirement::parse("14.0.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!("14.0.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_starts_with_v() {
        let result = Requirement::parse("v14.0.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!("14.0.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_semver_semantics() {
        let result = Requirement::parse(">= 12.0.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(">=12.0.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_pipe_statements() {
        let result = Requirement::parse("^12 || ^13 || ^14");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(
                ">=12.0.0 <13.0.0-0||>=13.0.0 <14.0.0-0||>=14.0.0 <15.0.0-0",
                format!("{}", reqs)
            );
        }
    }

    #[test]
    fn parse_handles_tilde_with_equals() {
        let result = Requirement::parse("~=14.4");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(">=14.4.0 <14.5.0-0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_tilde_with_equals_and_patch() {
        let result = Requirement::parse("~=14.4.3");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(">=14.4.3 <14.5.0-0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_v_within_string() {
        let result = Requirement::parse(">v15.5.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(">15.5.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_v_with_space() {
        let result = Requirement::parse(">= v10.0.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!(">=10.0.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_handles_equal_with_v() {
        let result = Requirement::parse("=v10.22.0");

        assert!(result.is_ok());
        if let Ok(reqs) = result {
            assert_eq!("10.22.0", format!("{}", reqs));
        }
    }

    #[test]
    fn parse_returns_error_for_invalid_reqs() {
        let result = Requirement::parse("12.%");
        println!("{:?}", result);

        assert!(result.is_err());
    }
}
