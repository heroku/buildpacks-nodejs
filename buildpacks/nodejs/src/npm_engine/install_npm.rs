use crate::NpmEngineBuildpack;
use crate::NpmEngineBuildpackError;
use bullet_stream::global::print;
use fun_run::{CommandWithName, NamedOutput};
use heroku_nodejs_utils::http::{get, ResponseExt};
use heroku_nodejs_utils::npmjs_org::PackagePackument;
use heroku_nodejs_utils::vrs::Version;
use libcnb::build::BuildContext;
use libcnb::data::layer_name;
use libcnb::layer::{
    CachedLayerDefinition, EmptyLayerCause, InvalidMetadataAction, LayerState, RestoredLayerAction,
};
use libherokubuildpack::tar::decompress_tarball;
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::path::{Path, PathBuf};
use std::process::Command;

pub(crate) fn install_npm(
    context: &BuildContext<NpmEngineBuildpack>,
    npm_packument: &PackagePackument,
    node_version: &Version,
) -> Result<(), libcnb::Error<NpmEngineBuildpackError>> {
    let npm_version = &npm_packument.version;

    let new_metadata = NpmEngineLayerMetadata {
        node_version: node_version.to_string(),
        npm_version: npm_version.to_string(),
        layer_version: LAYER_VERSION.to_string(),
        arch: context.target.arch.clone(),
        os: context.target.os.clone(),
    };

    let npm_engine_layer = context.cached_layer(
        layer_name!("npm_engine"),
        CachedLayerDefinition {
            build: true,
            launch: true,
            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
            restored_layer_action: &|old_metadata: &NpmEngineLayerMetadata, _| {
                if old_metadata == &new_metadata {
                    Ok((RestoredLayerAction::KeepLayer, vec![]))
                } else {
                    Ok((
                        RestoredLayerAction::DeleteLayer,
                        changed_metadata_fields(old_metadata, &new_metadata),
                    ))
                }
            },
        },
    )?;

    match npm_engine_layer.state {
        LayerState::Restored { .. } => {
            print::sub_bullet("Using cached npm");
        }
        LayerState::Empty { ref cause } => {
            npm_engine_layer.write_metadata(new_metadata)?;

            if let EmptyLayerCause::RestoredLayerAction { cause } = cause {
                print::sub_bullet(format!(
                    "Invalidating cached npm ({} changed)",
                    cause.join(", ")
                ));
            }

            let downloaded_package_path = npm_engine_layer.path().join("npm.tgz");
            let npm_cli_script = npm_engine_layer.path().join("package/bin/npm-cli.js");

            // this install process is generalized from the npm install script at:
            // https://www.npmjs.com/install.sh
            download_and_unpack_release(
                &npm_packument.dist.tarball,
                &downloaded_package_path,
                &npm_engine_layer.path(),
            )?;
            remove_existing_npm_installation(&npm_cli_script)?;
            install_npm_package(&npm_cli_script, &downloaded_package_path)?;
        }
    }

    Ok(())
}

fn download_and_unpack_release(
    download_from: &String,
    download_to: &Path,
    unpack_into: &Path,
) -> Result<(), NpmInstallError> {
    get(download_from)
        .call_sync()
        .and_then(|res| res.download_to_file_sync(download_to))
        .map_err(NpmInstallError::Download)
        .and_then(|()| {
            File::open(download_to)
                .map_err(|e| NpmInstallError::OpenTarball(download_to.to_path_buf(), e))
        })
        .and_then(|mut npm_tgz_file| {
            decompress_tarball(&mut npm_tgz_file, unpack_into)
                .map_err(|e| NpmInstallError::DecompressTarball(download_to.to_path_buf(), e))
        })?;
    Ok(())
}

fn remove_existing_npm_installation(npm_cli_script: &Path) -> Result<(), NpmInstallError> {
    print::sub_bullet("Removing npm bundled with Node.js");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "rm",
            "npm",
            "-gf",
            "--loglevel=silent",
        ])
        .named_output()
        .map_err(NpmInstallError::RemoveExistingNpmInstall)
        .map(|_| ())
}

fn install_npm_package(npm_cli_script: &Path, package: &Path) -> Result<(), NpmInstallError> {
    print::sub_bullet("Installing requested npm");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "install",
            "-gf",
            &package.to_string_lossy(),
        ])
        .named_output()
        .and_then(NamedOutput::nonzero_captured)
        .map_err(NpmInstallError::InstallNpm)
        .map(|_| ())
}

fn changed_metadata_fields(
    old: &NpmEngineLayerMetadata,
    new: &NpmEngineLayerMetadata,
) -> Vec<String> {
    let mut changed = vec![];
    if old.npm_version != new.npm_version {
        changed.push("npm version".to_string());
    }
    if old.node_version != new.node_version {
        changed.push("node version".to_string());
    }
    if old.layer_version != new.layer_version {
        changed.push("layer version".to_string());
    }
    if old.os != new.os {
        changed.push("operating system".to_string());
    }
    if old.arch != new.arch {
        changed.push("compute architecture".to_string());
    }
    changed.sort();
    changed
}

const LAYER_VERSION: &str = "1";

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub(crate) struct NpmEngineLayerMetadata {
    layer_version: String,
    npm_version: String,
    node_version: String,
    arch: String,
    os: String,
}

#[derive(Debug)]
pub(crate) enum NpmInstallError {
    Download(heroku_nodejs_utils::http::Error),
    OpenTarball(PathBuf, std::io::Error),
    DecompressTarball(PathBuf, std::io::Error),
    RemoveExistingNpmInstall(fun_run::CmdError),
    InstallNpm(fun_run::CmdError),
}

impl From<NpmInstallError> for libcnb::Error<NpmEngineBuildpackError> {
    fn from(value: NpmInstallError) -> Self {
        libcnb::Error::BuildpackError(NpmEngineBuildpackError::NpmInstall(value))
    }
}
