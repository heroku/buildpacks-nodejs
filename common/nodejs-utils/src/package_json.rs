use crate::vrs::{Requirement, Version};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use thiserror::Error;

#[derive(Deserialize, Debug, Clone)]
pub struct PackageJson {
    pub name: String,
    pub version: Option<Version>,
    pub engines: Option<Engines>,
    pub scripts: Option<Scripts>,
    pub main: Option<String>,
    pub dependencies: Option<HashMap<String, String>>,
    #[serde(rename = "devDependencies")]
    pub dev_dependencies: Option<HashMap<String, String>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Engines {
    pub node: Option<Requirement>,
    pub yarn: Option<Requirement>,
    pub npm: Option<Requirement>,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct Scripts {
    pub start: Option<String>,
    pub build: Option<String>,
    pub heroku_prebuild: Option<String>,
    pub heroku_build: Option<String>,
    pub heroku_postbuild: Option<String>,
}

#[derive(Error, Debug)]
pub enum PackageJsonError {
    #[error("Could not read package.json. {0}")]
    AccessError(std::io::Error),
    #[error("Could not parse package.json. {0}")]
    ParseError(serde_json::Error),
}

impl PackageJson {
    /// Reads a package.json from a path into a `PackageJson` struct.
    ///
    /// # Errors
    ///
    /// * Invalid/malformed JSON
    /// * Path does not exist or is unreadable
    /// * Version strings are invalid/malformed
    pub fn read<P: AsRef<Path>>(path: P) -> Result<Self, PackageJsonError> {
        let file = File::open(path).map_err(PackageJsonError::AccessError)?;
        let rdr = BufReader::new(file);
        serde_json::from_reader(rdr).map_err(PackageJsonError::ParseError)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::Builder;

    #[test]
    fn read_valid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{\"name\": \"foo\",\"version\": \"0.0.0\"}}").unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(pkg.name, "foo");
        assert_eq!(pkg.version.unwrap().to_string(), "0.0.0");
    }

    #[test]
    fn read_valid_package_with_node_engine() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(
            f,
            "{{
            \"name\": \"foo\",
            \"version\": \"0.0.0\",
            \"engines\": {{
                \"node\": \"16.0.0\"
            }}
        }}"
        )
        .unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(&pkg.engines.unwrap().node.unwrap().to_string(), "16.0.0");
    }

    #[test]
    fn read_missing_package() {
        let res = PackageJson::read(Path::new("/over/there/package.json"));
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not read package.json"));
    }

    #[test]
    fn read_invalid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{").unwrap();
        let res = PackageJson::read(f.path());
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not parse package.json"));
    }
}
