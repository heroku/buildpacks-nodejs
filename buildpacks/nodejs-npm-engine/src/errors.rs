use crate::install_npm::NpmEngineLayerError;
use crate::BUILDPACK_NAME;
use crate::{node, npm};
use bullet_stream::state::Bullet;
use bullet_stream::{style, Print};
use heroku_nodejs_utils::package_json::PackageJsonError;
use heroku_nodejs_utils::vrs::Requirement;
use indoc::formatdoc;
use std::fmt::Display;
use std::io::{stderr, Stderr};

const USE_DEBUG_INFORMATION_AND_RETRY_BUILD: &str = "\
Use the debug information above to troubleshoot and retry your build.";

const SUBMIT_AN_ISSUE: &str = "\
If the issue persists and you think you found a bug in the buildpack then reproduce the issue \
locally with a minimal example and open an issue in the buildpack's GitHub repository with the details.";

#[derive(Debug)]
pub(crate) enum NpmEngineBuildpackError {
    PackageJson(PackageJsonError),
    MissingNpmEngineRequirement,
    InventoryParse(toml::de::Error),
    NpmVersionResolve(Requirement),
    NpmEngineLayer(NpmEngineLayerError),
    NodeVersion(node::VersionError),
    NpmVersion(npm::VersionError),
}

pub(crate) fn on_error(error: libcnb::Error<NpmEngineBuildpackError>) {
    let logger = Print::new(stderr()).without_header();
    match error {
        libcnb::Error::BuildpackError(buildpack_error) => {
            on_buildpack_error(buildpack_error, logger);
        }
        framework_error => on_framework_error(&framework_error, logger),
    }
}

fn on_buildpack_error(error: NpmEngineBuildpackError, logger: Print<Bullet<Stderr>>) {
    match error {
        NpmEngineBuildpackError::PackageJson(e) => on_package_json_error(e, logger),
        NpmEngineBuildpackError::MissingNpmEngineRequirement => {
            on_missing_npm_engine_requirement_error(logger);
        }
        NpmEngineBuildpackError::InventoryParse(e) => on_inventory_parse_error(&e, logger),
        NpmEngineBuildpackError::NpmVersionResolve(requirement) => {
            on_npm_version_resolve_error(&requirement, logger);
        }
        NpmEngineBuildpackError::NpmEngineLayer(e) => on_npm_engine_layer_error(e, logger),
        NpmEngineBuildpackError::NodeVersion(e) => on_node_version_error(e, logger),
        NpmEngineBuildpackError::NpmVersion(e) => on_npm_version_error(e, logger),
    }
}

fn on_package_json_error(error: PackageJsonError, logger: Print<Bullet<Stderr>>) {
    match error {
        PackageJsonError::AccessError(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file can’t be read. 
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = style::value("package.json")});
        }
        PackageJsonError::ParseError(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file \
                    can’t be parsed. Ensure {npm_install} runs locally to check the formatting in your file. 
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = style::value("package.json"), npm_install = style::value("npm install") });
        }
    }
}

fn on_missing_npm_engine_requirement_error(logger: Print<Bullet<Stderr>>) {
    logger.error(formatdoc! {"
        Missing {engines_key} key in {package_json}.
        
        This buildpack requires the `engines.npm` key to determine which engine versions to install.

        Retry your build. 

        {SUBMIT_AN_ISSUE}
    ", engines_key = style::value("engines.npm"), package_json = style::value("package.json") });
}

fn on_inventory_parse_error(error: &toml::de::Error, logger: Print<Bullet<Stderr>>) {
    print_error_details(logger, &error).error(formatdoc! {"
            Failed to load available {npm} versions.

            An unexpected error occurred while loading the available {npm} versions.
        
            {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

            {SUBMIT_AN_ISSUE}
        ", npm = style::value("npm") });
}

fn on_npm_version_resolve_error(requirement: &Requirement, logger: Print<Bullet<Stderr>>) {
    logger.error(formatdoc! {"
            Error resolving requested {npm} version {requested_version}.
            
            Can’t find the `npm` version that matches the requested version declared in {package_json} ({requested_version}).
    
            Verify that the requested version range matches a published version of {npm} by checking \
            https://www.npmjs.com/package/npm?activeTab=versions or trying the following command:
            
                $ npm show 'npm@{requirement}' versions
    
            Update the {engines_key} field in {package_json} to a single version or version range that \
            includes a published {npm} version.
    
            {SUBMIT_AN_ISSUE}
        ",
        npm = style::value("npm"),
        requested_version = style::value(requirement.to_string()),
        package_json = style::value("package.json"),
        engines_key = style::value("engines.npm")
    });
}

fn on_npm_engine_layer_error(error: NpmEngineLayerError, logger: Print<Bullet<Stderr>>) {
    match error {
        NpmEngineLayerError::Download(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Failed to download {npm}.

                    An unexpected error occurred while downloading the {npm} package. This error can occur due to an unstable network connection.

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::OpenTarball(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while opening the downloaded {npm} package file. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::DecompressTarball(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    An unexpected error occurred while extracting the contents of the downloaded {npm} package file.

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::RemoveExistingNpmInstall(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while removing the existing {npm} installation. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::InstallNpm(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while installing the downloaded {npm} package. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
            ", npm = style::value("npm") });
        }
    }
}

fn on_node_version_error(error: node::VersionError, logger: Print<Bullet<Stderr>>) {
    match error {
        node::VersionError::Command(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to determine {node} version information.
    
                    An unexpected error occurred while executing {node_version}. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", node = style::value("Node"), node_version = style::value(e.name()) });
        }
        node::VersionError::Parse(stdout, e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to parse {node} version information.
        
                    An unexpected error occurred while parsing version information from {output}. 
                    
                    {SUBMIT_AN_ISSUE}
                ", node = style::value("Node"), output = style::value(stdout) });
        }
    }
}

fn on_npm_version_error(error: npm::VersionError, logger: Print<Bullet<Stderr>>) {
    match error {
        npm::VersionError::Command(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to determine {npm} version information.

                    An unexpected error occurred while executing {npm_version}.  
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}
        
                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm"), npm_version = style::value(e.name())});
        }
        npm::VersionError::Parse(stdout, e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to parse {npm} version information.
        
                    An unexpected error occurred while parsing version information from {output}. 
                    
                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm"), output = style::value(stdout) });
        }
    }
}

fn on_framework_error(
    error: &libcnb::Error<NpmEngineBuildpackError>,
    logger: Print<Bullet<Stderr>>,
) {
    print_error_details(logger, &error)
        .error(formatdoc! {"
            {buildpack_name} internal error.
    
            The framework used by this buildpack encountered an unexpected error.

            If you can't deploy to Heroku due to this issue, check the official Heroku Status page at \
            status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

            If the issue persists and you think you found a bug in the buildpack or framework, reproduce \
            the issue locally with a minimal example. Open an issue in the buildpack's GitHub repository \
            and include the details.

        ", buildpack_name = style::value(BUILDPACK_NAME) });
}

fn print_error_details(
    logger: Print<Bullet<Stderr>>,
    error: &impl Display,
) -> Print<Bullet<Stderr>> {
    logger
        .bullet(style::important("DEBUG INFO:"))
        .sub_bullet(error.to_string())
        .done()
}
