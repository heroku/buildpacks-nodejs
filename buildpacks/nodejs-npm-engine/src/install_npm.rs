use bullet_stream::state::SubBullet;
use bullet_stream::{style, Print};
use fun_run::{CommandWithName, NamedOutput};
use libcnb::build::BuildContext;
use libcnb::data::layer_name;
use libcnb::layer::{
    CachedLayerDefinition, EmptyLayerCause, InvalidMetadataAction, LayerState, RestoredLayerAction,
};
use libherokubuildpack::download::{download_file, DownloadError};
use libherokubuildpack::tar::decompress_tarball;
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::io::Stdout;
use std::path::Path;
use std::process::Command;

use heroku_nodejs_utils::inv::Release;
use heroku_nodejs_utils::vrs::Version;

use crate::errors::NpmEngineBuildpackError;
use crate::NpmEngineBuildpack;

pub(crate) fn install_npm(
    context: &BuildContext<NpmEngineBuildpack>,
    npm_release: &Release,
    node_version: &Version,
    mut logger: Print<SubBullet<Stdout>>,
) -> Result<Print<SubBullet<Stdout>>, libcnb::Error<NpmEngineBuildpackError>> {
    let new_metadata = NpmEngineLayerMetadata {
        node_version: node_version.to_string(),
        npm_version: npm_release.version.to_string(),
        layer_version: LAYER_VERSION.to_string(),
        arch: context.target.arch.clone(),
        os: context.target.os.clone(),
    };

    let npm_engine_layer = context.cached_layer(
        layer_name!("npm_engine"),
        CachedLayerDefinition {
            build: true,
            launch: true,
            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
            restored_layer_action: &|old_metadata: &NpmEngineLayerMetadata, _| {
                if old_metadata == &new_metadata {
                    Ok((RestoredLayerAction::KeepLayer, vec![]))
                } else {
                    Ok((
                        RestoredLayerAction::DeleteLayer,
                        changed_metadata_fields(old_metadata, &new_metadata),
                    ))
                }
            },
        },
    )?;

    match npm_engine_layer.state {
        LayerState::Restored { .. } => {
            logger = logger.sub_bullet("Using cached npm");
        }
        LayerState::Empty { ref cause } => {
            npm_engine_layer.write_metadata(new_metadata)?;

            if let EmptyLayerCause::RestoredLayerAction { cause } = cause {
                logger = logger.sub_bullet(format!(
                    "Invalidating cached npm ({} changed)",
                    cause.join(", ")
                ));
            }

            let downloaded_package_path = npm_engine_layer.path().join("npm.tgz");
            let npm_cli_script = npm_engine_layer.path().join("package/bin/npm-cli.js");

            // this install process is generalized from the npm install script at:
            // https://www.npmjs.com/install.sh
            logger = download_and_unpack_release(
                &npm_release.url,
                &downloaded_package_path,
                &npm_engine_layer.path(),
                logger,
            )?;
            logger = remove_existing_npm_installation(&npm_cli_script, logger)?;
            logger = install_npm_package(&npm_cli_script, &downloaded_package_path, logger)?;
        }
    }

    Ok(logger)
}

fn download_and_unpack_release(
    download_from: &String,
    download_to: &Path,
    unpack_into: &Path,
    logger: Print<SubBullet<Stdout>>,
) -> Result<Print<SubBullet<Stdout>>, NpmEngineLayerError> {
    let timer = logger.start_timer(format!("Downloading {}", style::value(download_from)));
    download_file(download_from, download_to)
        .map_err(NpmEngineLayerError::Download)
        .and_then(|()| File::open(download_to).map_err(NpmEngineLayerError::OpenTarball))
        .and_then(|mut npm_tgz_file| {
            decompress_tarball(&mut npm_tgz_file, unpack_into)
                .map_err(NpmEngineLayerError::DecompressTarball)
        })?;
    Ok(timer.done())
}

fn remove_existing_npm_installation(
    npm_cli_script: &Path,
    mut logger: Print<SubBullet<Stdout>>,
) -> Result<Print<SubBullet<Stdout>>, NpmEngineLayerError> {
    logger = logger.sub_bullet("Removing npm bundled with Node.js");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "rm",
            "npm",
            "-gf",
            "--loglevel=silent",
        ])
        .named_output()
        .map_err(NpmEngineLayerError::RemoveExistingNpmInstall)
        .map(|_| logger)
}

fn install_npm_package(
    npm_cli_script: &Path,
    package: &Path,
    mut logger: Print<SubBullet<Stdout>>,
) -> Result<Print<SubBullet<Stdout>>, NpmEngineLayerError> {
    logger = logger.sub_bullet("Installing requested npm");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "install",
            "-gf",
            &package.to_string_lossy(),
        ])
        .named_output()
        .and_then(NamedOutput::nonzero_captured)
        .map_err(NpmEngineLayerError::InstallNpm)
        .map(|_| logger)
}

fn changed_metadata_fields(
    old: &NpmEngineLayerMetadata,
    new: &NpmEngineLayerMetadata,
) -> Vec<String> {
    let mut changed = vec![];
    if old.npm_version != new.npm_version {
        changed.push("npm version".to_string());
    }
    if old.node_version != new.node_version {
        changed.push("node version".to_string());
    }
    if old.layer_version != new.layer_version {
        changed.push("layer version".to_string());
    }
    if old.os != new.os {
        changed.push("operating system".to_string());
    }
    if old.arch != new.arch {
        changed.push("compute architecture".to_string());
    }
    changed.sort();
    changed
}

const LAYER_VERSION: &str = "1";

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub(crate) struct NpmEngineLayerMetadata {
    layer_version: String,
    npm_version: String,
    node_version: String,
    arch: String,
    os: String,
}

#[derive(Debug)]
pub(crate) enum NpmEngineLayerError {
    Download(DownloadError),
    OpenTarball(std::io::Error),
    DecompressTarball(std::io::Error),
    RemoveExistingNpmInstall(fun_run::CmdError),
    InstallNpm(fun_run::CmdError),
}

impl From<NpmEngineLayerError> for libcnb::Error<NpmEngineBuildpackError> {
    fn from(value: NpmEngineLayerError) -> Self {
        libcnb::Error::BuildpackError(NpmEngineBuildpackError::NpmEngineLayer(value))
    }
}
