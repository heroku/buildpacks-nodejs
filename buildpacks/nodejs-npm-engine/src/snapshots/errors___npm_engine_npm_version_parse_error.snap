---
source: buildpacks/nodejs-npm-engine/src/errors.rs
---
- Debug Info:
  - Failed to parse version.

! Failed to parse npm version
!
! An unexpected error occurred while parsing npm version information from 'not.a.version'.
!
! The causes for this error are unknown. We do not have suggestions for diagnosis or a workaround at this time. You can help our understanding by sharing your buildpack log and a description of the issue at:
! https://github.com/heroku/buildpacks-nodejs/issues
!
! If you're able to reproduce the problem with an example application and the `pack` build tool (https://buildpacks.io/docs/for-platform-operators/how-to/integrate-ci/pack/), adding that information to the discussion will also help. Once we have more information around the causes of this error we may update this message.
